src/hooks/usePersistentDominantColorFromImage.ts

```diff
@@ -10,52 +12,86 @@ enum State {
+// settings a default value to easily check against it instead of just undefined and to make the value be always a string
+storage.set(DEFAULT_MMKV_KEY, NOT_DEFINED_MMKV_VALUE);
+
+const DEFAULT_COLOR_TO_MEASURE_AGAINST = '#333333';
+const DEFAULT_SIZE_FOR_COLOR_PROBING = 80;
 
 type Result = {
-  state: State;
   result: string | undefined;
+  state: State;
 };
 
+type ConfigOptions = {
+  colorToMeasureAgainst?: string;
+  signUrl?: boolean;
+};
 export default function usePersistentDominantColorFromImage(
-  url: string,
-  colorToMeasureAgainst: string = '#333333'
+  url: string | undefined | null,
+  options: ConfigOptions = {
+    colorToMeasureAgainst: DEFAULT_COLOR_TO_MEASURE_AGAINST,
+    signUrl: false,
+  }
 ): Result {
+  const previousUrl = usePrevious(url);
+  const { colorToMeasureAgainst: passedColor, signUrl } = options;
+  const externalUrl = useMemo(
+    () =>
+      signUrl && url
+        ? maybeSignUri(url, { w: DEFAULT_SIZE_FOR_COLOR_PROBING })
+        : url,
+    [signUrl, url]
+  );
+  const colorToMeasureAgainst = passedColor ?? DEFAULT_COLOR_TO_MEASURE_AGAINST;
   const [dominantColor, setPersistentDominantColor] = useMMKVString(
-    (url || '') as string,
+    url || DEFAULT_MMKV_KEY,
     storage
   );
-
+  // This is the most recent value saved in MMKV store, we can't use the value from the hook
+  // because it's updated a cycle behind due to onChange listener and useState usage
+  const currentStoreValue = storage.getString(url || DEFAULT_MMKV_KEY);
   const [state, setState] = useState<State>(
-    dominantColor ? State.loaded : url ? State.loading : State.init
+    currentStoreValue !== NOT_DEFINED_MMKV_VALUE ? State.loaded : State.init
   );
-  useEffect(() => {
-    if (!dominantColor) {
-      if (url) {
-        setState(State.loading);
-      } else {
-        setState(State.init);
-      }
-    }
-  }, [dominantColor, url]);
 
   useEffect(() => {
-    if ((state === State.loading || state === State.init) && url) {
+    if (
+      currentStoreValue !== NOT_DEFINED_MMKV_VALUE &&
+      currentStoreValue !== undefined
+    ) {
+      setState(State.loaded);
+    } else if (externalUrl && (state === State.init || url !== previousUrl)) {
       setState(State.loading);
-      getDominantColorFromImage(url, colorToMeasureAgainst)
+      getDominantColorFromImage(externalUrl, colorToMeasureAgainst)
         .then(color => {
           setPersistentDominantColor(color);
-        })
-        .finally(() => {
           setState(State.loaded);
+        })
+        .catch(() => {
+          setState(State.failed);
         });
     }
-  }, [colorToMeasureAgainst, setPersistentDominantColor, state, url]);
-
-  return {
-    result: dominantColor,
+  }, [
+    colorToMeasureAgainst,
+    externalUrl,
+    setPersistentDominantColor,
     state,
-  };
+    url,
+    previousUrl,
+  ]);
+
+  return useMemo(
+    () => ({
+      result:
+        dominantColor === NOT_DEFINED_MMKV_VALUE ? undefined : dominantColor,
```

Comment:
feel like we should have a fallback here, appleBlue would work. 

would rather always expect a color value vs having to program around type issues outside of this

Reply:
I disagree, we have a bunch of logic that depend on the fact that there can be a color form image or not. I would leave it as is with possibility of undefined, adding a fallback is as easy as using nullish coalescing operator `??` or logical OR `||`